src/main/java/redis/clients/jedis/ClientSideCache.java

```diff
@@ -1,71 +1,99 @@
+  private final ReentrantLock writeLock = new ReentrantLock();
 
-  public ClientSideCache() {
-    this.cache = new HashMap<>();
+  protected ClientSideCache() {
+    this.keyHashes = new ConcurrentHashMap<>();
   }
 
-  /**
-   * For testing purpose only.
-   * @param map 
-   */
-  ClientSideCache(Map<ByteBuffer, Object> map) {
-    this.cache = map;
-  }
+  public abstract void invalidateAll();
 
-  public final void clear() {
-    cache.clear();
-  }
+  protected abstract void invalidateAll(Iterable<Long> hashes);
 
-  public final void invalidateKeys(List list) {
+  final void invalidate(List list) {
     if (list == null) {
-      clear();
+      invalidateAll();
       return;
     }
 
-    list.forEach(this::invalidateKey);
+    list.forEach(this::invalidate0);
   }
 
-  private void invalidateKey(Object key) {
-    if (key instanceof byte[]) {
-      cache.remove(convertKey((byte[]) key));
-    } else {
-      throw new JedisException("" + key.getClass().getSimpleName() + " is not supported. Value: " + String.valueOf(key));
+  private void invalidate0(Object key) {
+    if (!(key instanceof byte[])) {
+      throw new AssertionError("" + key.getClass().getSimpleName() + " is not supported. Value: " + String.valueOf(key));
     }
-  }
 
-  protected void setKey(Object key, Object value) {
-    cache.put(getMapKey(key), value);
-  }
+    final ByteBuffer mapKey = makeKey((byte[]) key);
 
-  protected <T> T getValue(Object key) {
-    return (T) getMapValue(key);
+    Set<Long> hashes = keyHashes.get(mapKey);
+    if (hashes != null) {
+      writeLock.lock();
+      try {
+        invalidateAll(hashes);
+        keyHashes.remove(mapKey);
+      } finally {
+        writeLock.unlock();
+      }
+    }
   }
 
-  private Object getMapValue(Object key) {
-    return cache.get(getMapKey(key));
-  }
+  protected abstract void put(long hash, Object value);
+
+  protected abstract Object get(long hash);
 
-  private ByteBuffer getMapKey(Object key) {
-    if (key instanceof byte[]) {
-      return convertKey((byte[]) key);
-    } else {
-      return convertKey(SafeEncoder.encode(String.valueOf(key)));
+  final <T> T getValue(Function<CommandObject<T>, T> loader, CommandObject<T> command, String... keys) {
+
+    final long hash = getHash(command);
+
+    T value = (T) get(hash);
+    if (value != null) {
+      return value;
+    }
+
+    value = loader.apply(command);
+    if (value != null) {
+      writeLock.lock();
+      try {
+        put(hash, value);
+        for (String key : keys) {
+          ByteBuffer mapKey = makeKey(key);
+          if (keyHashes.containsKey(mapKey)) {
+            keyHashes.get(mapKey).add(hash);
+          } else {
+            Set<Long> set = new HashSet<>();
+            set.add(hash);
+            keyHashes.put(mapKey, set);
+          }
+        }
+      } finally {
+        writeLock.unlock();
+      }
     }
+
+    return value;
+  }
+
+  protected abstract long getHash(CommandObject command);
```

Comment:
Why have different ways of hashing a command? Also, why bind the hashing algorithm to the cache library that we use? Why not allow using Caffeine as a cache store with Guava's hashing routine? That would mean another interface for hash computation.

Reply:
> Why not allow using Caffeine as a cache store with Guava's hashing routine?

@gerzse This is exactly what I'm trying to achieve.

Guava has a nice hash solution where we can do `newHasher().put(a).put(b).put(c).hash()`.
But most other library only allows `function.hash(a)`. So we have to do `function.hash(array[function.hash(a), function.hash(b), function.hash(c)])`.
We can add an interface for later approach. But it makes the append style approach of Guava unusable (hashing is still possible in second way).
On the other hand, if we create an interface to support the first approach, it makes most hashing libraries unusable.
Now, what is best approach we can do?